lib/active_merchant/billing/gateways/redsys_rest.rb

```diff
@@ -0,0 +1,433 @@
+        requires!(options, :order_id)
+
+        post = {}
+        add_action(post, :purchase, options)
+        add_amount(post, money, options)
+        add_order(post, options[:order_id])
+        add_payment(post, payment)
+        add_description(post, options)
+        add_direct_payment(post, options)
+        add_threeds(post, options)
+
+        commit(post, options)
+      end
+
+      def authorize(money, payment, options = {})
+        requires!(options, :order_id)
+
+        post = {}
+        add_action(post, :authorize, options)
+        add_amount(post, money, options)
+        add_order(post, options[:order_id])
+        add_payment(post, payment)
+        add_description(post, options)
+        add_direct_payment(post, options)
+        add_threeds(post, options)
+
+        commit(post, options)
+      end
+
+      def capture(money, authorization, options = {})
+        post = {}
+        add_action(post, :capture)
+        add_amount(post, money, options)
+        order_id, = split_authorization(authorization)
+        add_order(post, order_id)
+        add_description(post, options)
+
+        commit(post, options)
+      end
+
+      def void(authorization, options = {})
+        requires!(options, :order_id)
+
+        post = {}
+        add_action(post, :cancel)
+        order_id, amount, currency = split_authorization(authorization)
+        add_amount(post, amount, currency: currency)
+        add_order(post, order_id)
+        add_description(post, options)
+
+        commit(post, options)
+      end
+
+      def refund(money, authorization, options = {})
+        requires!(options, :order_id)
+
+        post = {}
+        add_action(post, :refund)
+        add_amount(post, money, options)
+        order_id, = split_authorization(authorization)
+        add_order(post, order_id)
+        add_description(post, options)
+
+        commit(post, options)
+      end
+
+      def verify(creditcard, options = {})
+        requires!(options, :order_id)
+
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, creditcard, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((PAN\"=>\")(\d+)), '\1[FILTERED]').
+          gsub(%r((CVV2\"=>\")(\d+)), '\1[FILTERED]')
+      end
+
+      private
+
+      def add_direct_payment(post, options)
+        # Direct payment skips 3DS authentication. We should only apply this if execute_threed is false
+        # or authentication data is not present. Authentication data support to be added in the future.
+        return if options[:execute_threed] || options[:authentication_data]
+
+        post[:DS_MERCHANT_DIRECTPAYMENT] = true
+      end
+
+      def add_threeds(post, options)
+        post[:DS_MERCHANT_EMV3DS] = { threeDSInfo: 'CardData' } if options[:execute_threed]
```

Comment:
Since the remote tests say that 3DS isn't supported yet, what is this `add_threeds` function doing here?

Reply:
Great question! This provides functionality for the initial setup to go to the 3DS server or the payment server. The logic to have this was already in place so rather than scrap it, I kept it and will make a note of it when we get to adding 3DS support here (the following ticket).